### DISCUSSION TITLE
Expose external energy meters through mqtt

### DISCUSSION THREAD

**Discussion by jan--f** [jagheterfredrik/wallbox-mqtt-bridge#48] [Ideas]
Hi,
I have a Pulsar with a Carlo Gavazzi EM340 attached via modbus for load balancing. I want that data in HA and going through the wallbox seems the path of least resistance.
I think this could be useful generally for other people too. Folks who's HA already has an energy meter integrated can simply ignore these new sensor fields.
The redis instance already has the relevant data under the `m2w` key. Here is an excerpt from my box:
```
137) "Carlo3P_addr1.energy_wh.value"
138) "7.7789e+06"
139) "Carlo3P_addr1.line1.current.valid"
140) "1"
141) "Carlo3P_addr1.line2.current.valid"
142) "1"
143) "Carlo3P_addr1.line3.current.valid"
144) "1"
145) "Carlo3P_addr1.line1.current.value"
146) "5.372"
147) "Carlo3P_addr1.line2.current.value"
148) "5.148"
149) "Carlo3P_addr1.line3.current.value"
150) "5.265"
151) "Carlo3P_addr1.line1.voltage.value"
152) "231.1"
153) "Carlo3P_addr1.line2.voltage.value"
154) "231.8"
155) "Carlo3P_addr1.line3.voltage.value"
156) "231.1"
157) "Carlo3P_addr1.line1.power.value"
158) "1231.7"
159) "Carlo3P_addr1.line2.power.value"
160) "1190"
161) "Carlo3P_addr1.line3.power.value"
162) "1210.6"
163) "PBO.energy_wh.value"
164) "7.7789e+06"
165) "PBO.energy_wh.valid"
166) "1"
167) "PBO.line1.current.valid"
168) "1"
169) "PBO.line2.current.valid"
170) "1"
171) "PBO.line3.current.valid"
172) "1"
173) "PBO.line1.current.value"
174) "5.372"
175) "PBO.line2.current.value"
176) "5.144"
177) "PBO.line3.current.value"
178) "5.265"
179) "PBO.line1.voltage.value"
180) "231.1"
181) "PBO.line2.voltage.value"
182) "231.7"
183) "PBO.line3.voltage.value"
184) "231"
185) "PBO.line1.power.value"
186) "1231.9"
187) "PBO.line2.power.value"
188) "1188.7"
189) "PBO.line3.power.value"
190) "1210.3"
```

I'd be happy to contribute the code. However I'm not sure which field to use. I imagine the `PBO` and `Carlo3P_addr1` prefixes are the same device (I have only the one meter installed in my setup) and was wondering if I can consider `PBO` as the generic "external meter" key under which other meter models would show up to.
Maybe other know more and can hint to more relevant info?

I think the wallbox can also read two meters in a PV setup. Not sure though what that would look like but I assume that would be interesting data too.

**Comment by tronikos** [comment]
See https://community.home-assistant.io/t/wallbox-pulsar-plus-integration/200339/751
That just forwards the internal mqtt topic so it's faster and I assume less resource intensive than having the script in this repo poll redis and publish to mqtt.

You might also find https://community.home-assistant.io/t/energy-disaggregation-or-non-intrusive-load-monitoring-nilm-blueprint/619672 useful.

**Reply by jan--f** [reply]
Right that does work, thanks for the comment.
I still think it might be worth pursuing the addition to the mqtt-bridge. Unless I'm missing something the internal mosquitto instance does not expose the per phase power consumption (only current and voltage, I suppose that suffices but why not use the measured data) and the accumulated energy counter only updates every 5 minutes it seems.

Not sure if the added resource consumption is something to worry about in this context.

**Reply by tronikos** [reply]
Since I rely on its power boost to stay within my electrical panel capacity I worry about its responsiveness. If I remember right its CPU utilization is already constantly at 40% or somewhere around there.

You can easily create a template sensor to multiply voltage with current and the non frequent update of the accumulated energy counter isn't a problem for usage in the energy dashboard.

**Reply by jagheterfredrik** [reply]
That sounds insanely high, on my Wallbox the bridge consumes sub-1% CPU.

**Reply by tronikos** [reply]
That's the total across all processes. I understand the additional load will likely be relatively small but I'm cautious wasting CPU cycles in order to keep the device responsive for load shedding.

**Reply by tronikos** [reply]
I'm curious, what's your total utilization without the power meter?

**Reply by jan--f** [reply]
fwiw here's a snapshot from my box with power meter.

![2024-03-17-084756_1044x534](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/assets/458501/7f13c286-04ab-4d84-b2f2-b499436448c4)

**Reply by jan--f** [reply]
Gave #50 a quick test

![2024-03-17-090720_1059x531](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/assets/458501/86204b76-3be5-4337-9b1d-9365f49834c7)

Albeit this being poor measurement, at least in this sample of 1 the PR has no detectable impact.

**Reply by jagheterfredrik** [reply]
I did notice high memory consumption and CPU spikes, probably from gc. I hope they’re fixed with #51

**Reply by tronikos** [reply]
@jagheterfredrik since you don't have a power meter, I'm curious what does your CPU usage look like across all processes? If it's much less than us, it might be worth figuring out a way to stop writing all these power meter measurements into mysql without breaking power management functionality.

**Reply by jagheterfredrik** [reply]
<img width="626" alt="Screenshot 2024-03-18 at 10 41 53" src="https://github.com/jagheterfredrik/wallbox-mqtt-bridge/assets/253493/93bba366-4bc4-4b0f-b895-17fabda73cde">

It's not too different I suppose, the point is that the bridge is consuming next to nothing compared to m2w and mysql.

**Comment by jagheterfredrik** [comment]
Added in #50 thanks to @martonbtoth

**Comment by tronikos** [comment]
I just tried it. There is no noticeable resource impact. But there is a noticeable delay of up to 5 seconds. See

![image](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/assets/9987465/47c01582-69c2-4cf1-85b1-cbde3d139bb5)

Not sure why it's more than the polling interval of 1 sec. I guess the values on redis aren't updated as often as the ones in mqtt. I'll personally continue forwarding the internal mqtt topic via https://community.home-assistant.io/t/wallbox-pulsar-plus-integration/200339/751 since it's faster and likely more resource efficient.

**Reply by martonbtoth** [reply]
MQTT publishing is rate limited on the application side, the values in redis are updated much more frequently. I guess its the same frequency as new values are published to the internal mqtt broker.

**Reply by tronikos** [reply]
If you see the screenshot, the first increase of 1A was delayed by 5 seconds. A few other changes were also delayed by that much. I don't think it's due to the rate limiter. The rate limiter is configured at 0.2 threshold.

**Reply by jan--f** [reply]
I have been running this change since last night. I can't see any delays.

![2024-03-18-082310_1823x457](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/assets/458501/8068a529-2ac6-4890-bb67-cdc265cc06b7)

Every now and again the two values are slightly different though. Not too much to worry about I think but its curious nonetheless.

![2024-03-18-082337_1825x624](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/assets/458501/21feae31-3d56-47ff-8146-9786c6d0f98a)

**Reply by tronikos** [reply]
Can you compare them in home assistant history http://homeassistant.local:8123/history over a 5 minutes time window?

**Reply by jan--f** [reply]
Here is roughly the same time frame in home assistant.

![2024-03-18-102023_1649x549](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/assets/458501/cb42b2a9-78c2-4737-af9a-02e939af7e0f)
